LayoutTests/fast/attachment/resources/attachment-test-utils.js

```diff
@@ -0,0 +1,46 @@
+function hasThumbnailOrIcon(elem)
+{
+    if (window.internals) {
+        return internals.hasThumbnailOrIcon(elem);
+    }
+    return true;
+}
+
+function takeScreenshotWhenAttachmentsSettled(message, iteration = 0)
+{
+    console.assert(document.documentElement.classList.contains("reftest-wait"));
+
+    if (document.readyState != "complete")
+        return addEventListener("load", () => takeScreenshotWhenAttachmentsSettled(message, 0), { once: true });
+
+    let attachments = [...document.getElementsByTagName("attachment")];
+    console.assert(attachments.length);
+
+    let states = attachments.map(e => [e, hasThumbnailOrIcon(e)]);
+
+    if (states.every(s => s[1]))
+        return document.documentElement.classList.remove("reftest-wait");
+
+    let tooManyIterations = iteration >= (100 + attachments.length);
+    if (tooManyIterations) {
+        for (const [e, settled] of states) {
+            if (!settled) {
+                e.insertAdjacentText("afterend", "- no icon nor thumbnail (" + message + ")");
+            }
+        }
+
+        return document.documentElement.classList.remove("reftest-wait");
+    }
+
+    requestAnimationFrame(() => takeScreenshotWhenAttachmentsSettled(message, iteration + 1));
```

Comment:
How difficult do you think it would be to make this event-based, rather than polling for the icon?

Reply:
Good idea, I'll have a look...

Reply:
What did you make me do? 😄
I've removed the polling code and functions, now it's all event-driven.